File: src/contests/cg/puzzle/dart/max-area.dart
// 🎮 CodinGame Puzzle - max-area
// https://www.codingame.com/training/easy/max-area

import 'dart:io';

void main() {
  stdin.readLineSync();
  List<int> a = stdin.readLineSync()!.trim().split(' ').map(int.parse).toList();
  int left = 0, right = a.length - 1, best = 0;
  while (left < right) {
    int h = a[left] < a[right] ? a[left] : a[right];
    int area = h * (right - left);
    if (area > best) best = area;
    if (a[left] < a[right]) {
      left++;
    } else {
      right--;
    }
  }
  print(best);
}

File: src/contests/cg/puzzle/rust/max-area.rs
// 🎮 CodinGame Puzzle - max-area
// https://www.codingame.com/training/easy/max-area

fn rl() -> String {
    let mut s = String::new();
    std::io::stdin().read_line(&mut s).unwrap();
    s.trim().to_string()
}

fn main() {
    let _ = rl();
    let a: Vec<i64> = rl().split_whitespace().map(|x| x.parse().unwrap()).collect();
    let mut left = 0usize;
    let mut right = a.len() - 1;
    let mut best = 0i64;
    while left < right {
        let h = a[left].min(a[right]);
        let area = h * (right - left) as i64;
        if area > best {
            best = area;
        }
        if a[left] < a[right] {
            left += 1;
        } else {
            right -= 1;
        }
    }
    println!("{}", best);
}
